### PR TITLE
Add contract function and tests [solves #41]

### PR DESCRIPTION
**Pull request by dantop114**
Function implemented in `AbstractConference.sol`, comments explains the inner working.
Tests implemented!

### CODE REVIEW

**Comment by makoto** [comment]
There seems 27 failing tests https://travis-ci.org/wearekickback/contracts/jobs/619737699#L1330
Can you fix them?

File: contracts/AbstractConference.sol [hunk 1]
```diff
@@ -86,17 +89,53 @@ contract AbstractConference is Conference, GroupAdmin {
      * @dev Withdraws deposit after the event is over.
      */
     function withdraw() external onlyEnded {
-        require(payoutAmount > 0, 'payout is 0');
+        require(payoutAmount > 0, 'payout is 0');
```
Comment (makoto) [inline contracts/AbstractConference.sol]: Please do not put whitespaces

File: contracts/AbstractConference.sol [hunk 2]
```diff
@@ -86,17 +89,53 @@ contract AbstractConference is Conference, GroupAdmin {
...
      */
     function withdraw() external onlyEnded {
-        require(payoutAmount > 0, 'payout is 0');
+        require(payoutAmount > 0, 'payout is 0');        
         Participant storage participant = participants[msg.sender];
         require(participant.addr == msg.sender, 'forbidden access');
         require(cancelled || isAttended(msg.sender), 'event still active or you did not attend');
         require(participant.paid == false, 'already withdrawn');
-
+
```
Comment (makoto) [inline contracts/AbstractConference.sol]: Please do not put whitespaces

File: contracts/AbstractConference.sol [hunk 3]
```diff
@@ -86,17 +89,53 @@ contract AbstractConference is Conference, GroupAdmin {
...
+    * addresses[i] will receive values[i]
+    * The function emits the event SendAndWithdraw with the following informations:
+    * (addresses, values, participant address, payoutAmount - sum(values), payoutAmount)
+    */
+    function sendAndWithdraw(address payable[] calldata addresses, uint256[] calldata values) external onlyEnded {
+        require(payoutAmount > 0, 'payout is 0');        
+        Participant storage participant = participants[msg.sender];
+        require(participant.addr == msg.sender, 'forbidden access');
+        require(cancelled || isAttended(msg.sender), 'event still active or you did not attend');
+        require(participant.paid == false, 'already withdrawn');
```
Comment (makoto) [inline contracts/AbstractConference.sol]: l114-l118 is redundant to `withdraw` logic. What about extracting that as modifier for reuse?

File: contracts/AbstractConference.sol [hunk 4]
```diff
@@ -86,17 +89,53 @@ contract AbstractConference is Conference, GroupAdmin {
...
+        require(sumOfValues <= payoutAmount, 'payout amount is less than sum of values');
+        require(addresses.length == values.length, 'more addresses than values or viceversa');
+
+        participant.paid = true;
+
+        for(uint i = 0; i < addresses.length; i++) {
+            doWithdraw(addresses[i], values[i]);
+        }
+
+        uint256 amountLeft = payoutAmount - sumOfValues;
```
Comment (makoto) [inline contracts/AbstractConference.sol]: This is more of question. You pulled in safemath and used `.add`. Should we use `.sub` here or better to keep as `-`?

File: test/behaviors/conference.behavior.js [hunk 5]
```diff
@@ -466,6 +466,39 @@ function shouldBehaveLikeConference () {
...
+      await conference.sendAndWithdraw([accounts[2]], [donation, donation], {from:registered}).should.be.rejected;
+      assertBalanceWithDeposit((await getBalance(conference.address)), mulBN(deposit, 2))
+    })
+
+    it('payout amount is less than sum of values', async function(){
+      await conference.cancel({from:owner});
+      await conference.sendAndWithdraw([accounts[2]], [toWei('1', "ether")], {from:registered}).should.be.rejected;
+      assertBalanceWithDeposit((await getBalance(conference.address)), mulBN(deposit, 2))
+    })
+  })
```
Comment (makoto) [inline test/behaviors/conference.behavior.js:532]: Do we need a test case for `conference.sendAndWithdraw([], [])` as well?

File: contracts/zeppelin/SafeMath.sol [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+     * Requirements:
+     * - The divisor cannot be zero.
+     *
+     * _Available since v2.4.0._
+     */
+    function mod(uint256 a, uint256 b, string memory errorMessage) internal pure returns (uint256) {
+        require(b != 0, errorMessage);
+        return a % b;
+    }
+}
```
Comment (makoto) [inline contracts/zeppelin/SafeMath.sol]: No new line at the end of file

File: contracts/AbstractConference.sol [hunk 7]
```diff
@@ -86,17 +89,53 @@ contract AbstractConference is Conference, GroupAdmin {
...
         require(participant.paid == false, 'already withdrawn');
-
+        
         participant.paid = true;
         doWithdraw(msg.sender, payoutAmount);
         emit WithdrawEvent(msg.sender, payoutAmount);
     }
 
+    /**
+    * @dev sendAndWithDraw function allows to split _payoutAmount_ among addresses in _addresses_.
```
Comment (makoto) [inline contracts/AbstractConference.sol:107]: Good comments 👍

File: contracts/AbstractConference.sol (hunk 4 above)
Comment (dantop114) [reply contracts/AbstractConference.sol]: In l125 `sumOfValues` value is checked to be less than `payoutAmount` so I think we can keep it like this.

**Comment by dantop114** [comment]
Okay, all suggestions present in last commit (https://github.com/wearekickback/contracts/pull/43/commits/d94759f1c4077e70f23016d468cb3dcc0bad432e).
Anyway I can't work out why coverage tests are failing. Will work on this too.
**EDIT** Coverage tests are now working.

**Review by makoto — DISMISSED** [review]

File: contracts/AbstractConference.sol [hunk 8]
```diff
@@ -85,18 +97,42 @@ contract AbstractConference is Conference, GroupAdmin {
...
+        require(sumOfValues <= payoutAmount, 'payout amount is less than sum of values');
+        require(addresses.length == values.length, 'more addresses than values or viceversa');
+
+        participants[msg.sender].paid = true;
+
+        for(uint i = 0; i < addresses.length; i++) {
+            doWithdraw(addresses[i], values[i]);
+        }
+
+        uint256 amountLeft = payoutAmount - sumOfValues;
```
Comment (hiddentao) [inline contracts/AbstractConference.sol]: Should use `SafeMath`. In fact, all calculations within the contract should now use it for consistency sake.

File: contracts/AbstractConference.sol [hunk 9]
```diff
@@ -85,18 +97,42 @@ contract AbstractConference is Conference, GroupAdmin {
...
+    * _addresses_ contains ethereum addresses
+    * _values_ contains the value of eth/dai to give to addresses
+    * 
+    * addresses[i] will receive values[i]
+    * The function emits the event SendAndWithdraw with the following informations:
+    * (addresses, values, participant address, payoutAmount - sum(values), payoutAmount)
+    */
+    function sendAndWithdraw(address payable[] calldata addresses, uint256[] calldata values) external canWithdraw onlyEnded {
+        uint256 sumOfValues = 0;
+        for(uint i = 0; i < values.length; i++) {
```
Comment (hiddentao) [inline contracts/AbstractConference.sol]: This loop is redundant. Can do this check in the loop below and revert afterwards. Everything will still work as expected without any funds being withdrawn.

File: test/behaviors/conference.behavior.js [hunk 10]
```diff
@@ -466,6 +466,52 @@ function shouldBehaveLikeConference () {
...
+      assertBalanceWithDeposit((await getBalance(conference.address)), mulBN(deposit, 2))
+    })
+
+    it('payout amount is less than sum of values', async function(){
+      await conference.cancel({from:owner});
+      await conference.sendAndWithdraw([accounts[2]], [toWei('1', "ether")], {from:registered}).should.be.rejected;
+      assertBalanceWithDeposit((await getBalance(conference.address)), mulBN(deposit, 2))
+    })
+
+    it('send and withdraw with empty arrays should send funds to sender', async function(){
```
Comment (hiddentao) [inline test/behaviors/conference.behavior.js:502]: Need a test case for array with 2 items just to test that split calculations work properly.

**Review by hiddentao — CHANGES_REQUESTED** [review]

File: contracts/AbstractConference.sol (hunk 8 above)
Comment (dantop114) [reply contracts/AbstractConference.sol]: Okay done! Now the whole contract uses `SafeMath`.

File: contracts/AbstractConference.sol (hunk 9 above)
Comment (dantop114) [reply contracts/AbstractConference.sol]: Done in the last commit.

File: test/behaviors/conference.behavior.js (hunk 10 above)
Comment (dantop114) [reply test/behaviors/conference.behavior.js:502]: Added the test! Thank you for reviewing!

**Review by hiddentao — APPROVED** [review]

**Review by hiddentao — CHANGES_REQUESTED** [review]
Actually, one more change. The "2 array items" test should test with 2 _different_ values instead of the same value.

**Comment by dantop114** [comment]
> Actually, one more change. The "2 array items" test should test with 2 _different_ values instead of the same value.

Okay, done!

**Review by hiddentao — APPROVED** [review]

**Pull request closed without merging**